### ISSUE TITLE
Convert cwl to json input

### ISSUE THREAD

**Issue by dridk** [common-workflow-language/cwltool#384]
How can I generate my job.json file ? 
would be greate to have : 

cwltools make jobtemplate pipleline.cwl  > pipeline-job.json

**Comment by mr-c** [comment]
Hello @dridk , this is a great proposal -- I agree that would be very useful!

**Comment by denis-yuen** [comment]
@dridk FYI, this isn't part of cwltool although it is built on top of it. 
We have a similar function as part of Dockstore (see the "make a runtime json" part of https://dockstore.org/containers/quay.io/wtsicgp/dockstore-cgpmap ) from this part of the code https://github.com/common-workflow-language/cwlavro/blob/2.0.0/cwlavro-tools/src/main/java/io/cwl/avro/CWL.java#L122

**Comment by bolton-lab** [comment]
@dridk I am confused about how to setup the tool?  The documentation is good but it's still hard to follow the requirements.  I have a CWL file here https://github.com/kbolton-lab/analysis-workflows/blob/master/definitions/tools/bcftools_filter.cwl and added `CWL Path: /definitions/tools/bcftools_filter.cwl` and it uses a docker image that I did not create so I don't have the Dockerfile in my git repo.  Should we just be able to add that docker image as `Docker Image: hub.docker.com/mgibio/bcftools-cwl` add the 1.9 tag and run:

`dockstore tool convert entry2json --entry hub.docker.com/mgibio/bcftools-cwl:1.9`

**Comment by mr-c** [comment]
For `cwltool` it would be `cwltool --make-template https://github.com/kbolton-lab/analysis-workflows/raw/master/definitions/tools/bcftools_filter.cwl`

@bolton-lab  For asking questions about Dockstore, probably better to do that in https://github.com/dockstore/dockstore/issues

``` yaml
vcf:  # type "File"
    class: File
    path: a/file/path
output_vcf_name: "bcftools_filter.vcf.gz"  # default value of type "string". (optional)
output_type: "z"  # enum; valid values: "https://github.com/kbolton-lab/analysis-workflows/raw/master/definitions/tools/bcftools_filter.cwl#output_type/b", "https://github.com/kbolton-lab/analysis-workflows/raw/master/definitions/tools/bcftools_filter.cwl#output_type/u", "https://github.com/kbolton-lab/analysis-workflows/raw/master/definitions/tools/bcftools_filter.cwl#output_type/z", "https://github.com/kbolton-lab/analysis-workflows/raw/master/definitions/tools/bcftools_filter.cwl#output_type/v"
filter_string: a_string  # type "string"
filter_flag: "include"  # enum; valid values: "https://github.com/kbolton-lab/analysis-workflows/raw/master/definitions/tools/bcftools_filter.cwl#filter_flag/include", "https://github.com/kbolton-lab/analysis-workflows/raw/master/definitions/tools/bcftools_filter.cwl#filter_flag/exclude"
```

**Comment by denis-yuen** [comment]
@bolton-lab 
That command works on local files, so 
```
$ dockstore tool convert cwl2yaml --cwl bcftools_filter.cwl 
output_type: z
output_vcf_name: bcftools_filter.vcf.gz
filtered_vcf:
  path: /tmp/fill_me_in.txt
  class: File
filter_string: fill me in
vcf:
  path: /tmp/fill_me_in.txt
  class: File
filter_flag: include

$ dockstore tool convert cwl2json --cwl bcftools_filter.cwl 
{
  "output_type": "z",
  "output_vcf_name": "bcftools_filter.vcf.gz",
  "filtered_vcf": {
    "path": "/tmp/fill_me_in.txt",
    "class": "File"
  },
  "filter_string": "fill me in",
  "vcf": {
    "path": "/tmp/fill_me_in.txt",
    "class": "File"
  },
  "filter_flag": "include"
}
```

That said, the `make-template` command probably does what you need. I think we had logged a ticket to look at simplifying https://github.com/dockstore/dockstore/issues/2910

We should discuss further on Dockstore if needed.

Thanks @mr-c !

**Comment by bolton-lab** [comment]
Thanks for both of your suggestions.  I was only interesting in making a JSON input because DNAnexus is starting use this in their applets but it’s pretty odd and not really the same and they are just starting to support so maybe eventually they will get it be simple JSON structure other than their files which works pretty good.

**Comment by mr-c** [comment]
> I was only interesting in making a JSON input because DNAnexus

Do they not accept the YAML version? If that is unavoidable then I can add a JSON variant of `cwltool --make-template`
